GCC 6.1 build failure

* gcc 6.1.1

Possibly related: http://developers.redhat.com/blog/2016/02/29/why-cstdlib-is-more-complicated-than-you-might-think/

> For GCC 6 this has been fixed, by making the C++ library install its own version of math.h, which includes cmath, which then includes the real math.h from the C 
library.

*(Edit: Removed <> from quote because lolmarkdown)*

```
In file included from /usr/include/llvm/Support/DataTypes.h:35:0,
                 from /usr/include/llvm-c/Support.h:17,
                 from /usr/include/llvm-c/Core.h:18,
                 from src/libponyc/codegen/gendebug.h:5,
                 from src/libponyc/codegen/gendebug.cc:1:
/usr/include/c++/6.1.1/cmath:45:23: fatal error: math.h: No such file or directory
 #include_next <math.h>
                       ^
compilation terminated.
Makefile:445: recipe for target 'build/release/obj/libponyc/codegen/gendebug.o' failed
make: *** [build/release/obj/libponyc/codegen/gendebug.o] Error 1
make: *** Waiting for unfinished jobs....
```

Looks like I have this same issue following an update. Arch Linux (bog-standard), llvm 3.7.1, gcc 6.1.1 20160501 (not sure what version I had previously).

Doing `make verbose=t` produces

```
...
gendebug.cc
g++ -MMD -MP -march=native -Werror -Wconversion -Wno-sign-conversion -Wextra -Wall -mcx16 -g -DDEBUG -std=gnu++11 -fno-rtti  -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -c -o build/debug/obj/libponyc/codegen/gendebug.o src/libponyc/codegen/gendebug.cc  -I src/common/ -I src/libponyrt/ -isystem /usr/include/
In file included from /usr/include/llvm/Support/DataTypes.h:35:0,
                 from /usr/include/llvm-c/Support.h:17,
                 from /usr/include/llvm-c/Core.h:18,
                 from src/libponyc/codegen/gendebug.h:5,
                 from src/libponyc/codegen/gendebug.cc:1:
/usr/include/c++/6.1.1/cmath:45:23: fatal error: math.h: No such file or directory
 #include_next <math.h>
                       ^
compilation terminated.
Makefile:445: recipe for target 'build/debug/obj/libponyc/codegen/gendebug.o' failed
make: *** [build/debug/obj/libponyc/codegen/gendebug.o] Error 1
```

Fiddling around with the command line suggests the issue involves the `-isystem /usr/include/` argument; removing it allows the build to continue. The following change in Makefile (replacing -isystem with -I) allows the build to succeed. Tests run satisfactorily.

```
diff --git a/Makefile b/Makefile
index 364f23c..ebde1e1 100644
--- a/Makefile
+++ b/Makefile
@@ -231,7 +231,8 @@ endif
 # (3) a list of include directories for a set of libraries
 # (4) a list of the libraries to link against
 llvm.ldflags := $(shell $(LLVM_CONFIG) --ldflags)
-llvm.include := -isystem $(shell $(LLVM_CONFIG) --includedir)
+# llvm.include := -isystem $(shell $(LLVM_CONFIG) --includedir)
+llvm.include := -I $(shell $(LLVM_CONFIG) --includedir)
 llvm.libs    := $(shell $(LLVM_CONFIG) --libs) -lz -lncurses
 
 ifeq ($(OSTYPE), freebsd)
```

The search path with `-isystem /usr/include` is

```
ignoring nonexistent directory "/usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../x86_64-pc-linux-gnu/include"
ignoring duplicate directory "/usr/include"
#include "..." search starts here:
#include <...> search starts here:
 src/common/
 src/libponyrt/
 /usr/include/
 /usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../include/c++/6.1.1
 /usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../include/c++/6.1.1/x86_64-pc-linux-gnu
 /usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../include/c++/6.1.1/backward
 /usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/include
 /usr/local/include
 /usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/include-fixed
End of search list.
```

The search path with `-I /usr/include` is

```
ignoring nonexistent directory "/usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../x86_64-pc-linux-gnu/include"
ignoring duplicate directory "/usr/include/"
  as it is a non-system directory that duplicates a system directory
#include "..." search starts here:
#include <...> search starts here:
 src/common/
 src/libponyrt/
 /usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../include/c++/6.1.1
 /usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../include/c++/6.1.1/x86_64-pc-linux-gnu
 /usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/../../../../include/c++/6.1.1/backward
 /usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/include
 /usr/local/include
 /usr/lib/gcc/x86_64-pc-linux-gnu/6.1.1/include-fixed
 /usr/include
End of search list.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -25,6 +25,7 @@ All notable changes to the Pony compiler and standard library will be documented
 - Simplify contains() method on HashMap.
 - Lambda captures use the alias of the expression type.
 - Trace boxed primitives in union types.
+- Use -isystem for LLVM include directory only if it is not in search path.
 
 ### Added
 

--- a/Makefile
+++ b/Makefile
@@ -231,7 +231,17 @@ endif
 # (3) a list of include directories for a set of libraries
 # (4) a list of the libraries to link against
 llvm.ldflags := $(shell $(LLVM_CONFIG) --ldflags)
-llvm.include := -isystem $(shell $(LLVM_CONFIG) --includedir)
+llvm.include.dir := $(shell $(LLVM_CONFIG) --includedir)
+include.paths := $(shell echo | cc -v -E - 2>&1)
+ifeq (,$(findstring $(llvm.include.dir),$(include.paths)))
+# LLVM include directory is not in the existing paths;
+# put it at the top of the system list
+llvm.include := -isystem $(llvm.include.dir)
+else
+# LLVM include directory is already on the existing paths;
+# do nothing
+llvm.include :=
+endif
 llvm.libs    := $(shell $(LLVM_CONFIG) --libs) -lz -lncurses
 
 ifeq ($(OSTYPE), freebsd)

